src/back/auth/src/service/tokens_oidc.cpp

```diff
@@ -25,6 +29,7 @@ struct jwt_impl {
 
 	Verifier* GetVerifier(const std::string& keyId, const std::string& algoName)
```

Comment:
Тут мы не можем возвращать указатель, т.к. объект может быть удалён после того как мы его вернули.
```suggestion
	std::shared_ptr<Verifier> GetVerifier(const std::string& keyId, const std::string& algoName)
```

там ниже гже auto* verifier = _jwt->GetVerifier.... 
нужно убрать звёздочку
auto verifier = _jwt->GetVerifier

Reply:
А вот это не получилось. Ругается внутри GetVerifier в двух местах, поправить пока не получилось.
![image](https://github.com/user-attachments/assets/5da0366a-d2e4-4a98-9f98-ea9feda47c77)
![image](https://github.com/user-attachments/assets/23536fb3-539b-4694-a11d-212db6a9b17b)

```
user@pc:~/svetit$ make run-bin-auth
pipeline/schemas/generate.sh
make -C src/back/auth build-debug
make[1]: вход в каталог «/home/user/svetit/src/back/auth»
cmake -B build_debug -DUSERVER_SANITIZE='ub' -DCMAKE_BUILD_TYPE=Debug -DCMAKE_EXPORT_COMPILE_COMMANDS=ON
CMake Warning (dev) at CMakeLists.txt:8 (set):
  implicitly converting 'INT' to 'STRING' type.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Using userver framework from third_party/userver
-- Building utest with gtest and ubench with gbench
-- Building userver as a subproject
-- rseq-based acceleration is enabled by default
-- C compiler: /usr/bin/cc
-- C++ compiler: /usr/bin/c++
-- C++ standard 23
-- Using the default system linker, probably GNU ld
-- LTO: disabled (user request)
-- ccache: enabled
-- Found Git: /usr/bin/git  
-- Python: python3
-- Generating cmake files ...
-- Setting up the venv at /home/user/svetit/src/back/auth/build_debug/third_party/userver/venv-userver-cmake
-- Sanitizers are ON: ub
-- Found Boost: /usr/lib/x86_64-linux-gnu/cmake/Boost-1.74.0/BoostConfig.cmake (found version "1.74.0") found components: program_options filesystem regex stacktrace_basic stacktrace_backtrace 
-- boost: 1.74.0
-- CryptoPP include directories: /usr/include
-- CryptoPP libraries: /usr/lib/x86_64-linux-gnu/libcryptopp.so
-- libyamlcpp include directories: /usr/include
-- libyamlcpp libraries: /usr/lib/x86_64-linux-gnu/libyaml-cpp.so
-- fmt include directories: /usr/include
-- fmt libraries: /usr/lib/x86_64-linux-gnu/libfmt.so
-- cctz include directories: /usr/include
-- cctz libraries: /usr/lib/x86_64-linux-gnu/libcctz.so
-- DWCAS: Using boost::atomic
-- DWCAS works
-- Checking -ftemplate-backtrace-limit=0 - found
-- Checking -Wdisabled-optimization - found
-- Checking -Winvalid-pch - found
-- Checking -Wimplicit-fallthrough - found
-- Checking -Wlogical-op - found
-- Checking -Wformat=2 - found
-- Checking -Wno-error=deprecated-declarations - found
-- Checking -Wno-useless-cast - found
-- Checking -Wno-gnu-zero-variadic-macro-arguments - found
-- Checking -fmacro-prefix-map=a=b - found
-- Found Git: /usr/bin/git;/usr/bin/git  
-- Failed to retrieve git short hash
-- Putting userver into namespace 'userver': namespace userver { inline namespace v2_1_rc {  } }
-- UserverGTest include directories: /usr/include;/usr/include
-- UserverGTest libraries: /usr/lib/x86_64-linux-gnu/libgtest.a;/usr/lib/x86_64-linux-gnu/libgtest_main.a;/usr/lib/x86_64-linux-gnu/libgmock.a
-- UserverGBench include directories: /usr/include
-- UserverGBench libraries: /usr/lib/x86_64-linux-gnu/libbenchmark_main.a;/usr/lib/x86_64-linux-gnu/libbenchmark.so
-- Failed to retrieve git short hash
-- Userver version 2.1-rc
-- Found Boost: /usr/lib/x86_64-linux-gnu/cmake/Boost-1.74.0/BoostConfig.cmake (found version "1.74.0") found components: program_options filesystem locale regex iostreams 
-- c-ares include directories: /usr/include
-- c-ares libraries: /usr/lib/x86_64-linux-gnu/libcares.so
-- Nghttp2 include directories: /usr/include
-- Nghttp2 libraries: /usr/lib/x86_64-linux-gnu/libnghttp2.so
-- LibEv include directories: /usr/include
-- LibEv libraries: /usr/lib/x86_64-linux-gnu/libev.so
-- Context impl: ucontext
-- Sanitizers are ON: ub
-- Found Boost: /usr/lib/x86_64-linux-gnu/cmake/Boost-1.74.0/BoostConfig.cmake (found version "1.74.0")  
-- Found Boost: /usr/lib/x86_64-linux-gnu/cmake/Boost-1.74.0/BoostConfig.cmake (found version "1.74.0") found components: regex 
-- PostgreSQLInternal include directories: /usr/include/postgresql/16/server;/usr/include/postgresql/internal;/usr/include/postgresql
-- PostgreSQLInternal libraries: /usr/lib/x86_64-linux-gnu/libpq.a;/usr/lib/postgresql/16/lib/libpgcommon.a;/usr/lib/postgresql/16/lib/libpgport.a
-- GssApi include directories: /usr/include/gssapi
-- GssApi libraries: /usr/lib/x86_64-linux-gnu/libgssapi_krb5.so
-- C compiler: /usr/bin/cc
-- C++ compiler: /usr/bin/c++
-- C++ standard 23
-- Using the default system linker, probably GNU ld
-- LTO: disabled (user request)
-- ccache: enabled
-- Setting up the venv at /home/user/svetit/src/back/auth/build_debug/venv-testsuite-svetit_auth
-- Configuring done
-- Generating done
-- Build files have been written to: /home/user/svetit/src/back/auth/build_debug
cmake --build build_debug -j 12 --target svetit_auth
gmake[2]: вход в каталог «/home/user/svetit/src/back/auth/build_debug»
gmake[3]: вход в каталог «/home/user/svetit/src/back/auth/build_debug»
gmake[4]: вход в каталог «/home/user/svetit/src/back/auth/build_debug»
gmake[5]: вход в каталог «/home/user/svetit/src/back/auth/build_debug»
gmake[5]: вход в каталог «/home/user/svetit/src/back/auth/build_debug»
gmake[5]: вход в каталог «/home/user/svetit/src/back/auth/build_debug»
gmake[5]: вход в каталог «/home/user/svetit/src/back/auth/build_debug»
gmake[5]: вход в каталог «/home/user/svetit/src/back/auth/build_debug»
gmake[5]: вход в каталог «/home/user/svetit/src/back/auth/build_debug»
gmake[5]: вход в каталог «/home/user/svetit/src/back/auth/build_debug»
Consolidate compiler generated dependencies of target userver-pq-extra
gmake[5]: выход из каталога «/home/user/svetit/src/back/auth/build_debug»
Consolidate compiler generated dependencies of target userver-librseq
Scanning dependencies of target userver-uboost-coro
Consolidate compiler generated dependencies of target userver-http-parser
Consolidate compiler generated dependencies of target userver-llhttp
gmake[5]: выход из каталога «/home/user/svetit/src/back/auth/build_debug»
gmake[5]: выход из каталога «/home/user/svetit/src/back/auth/build_debug»
gmake[5]: выход из каталога «/home/user/svetit/src/back/auth/build_debug»
Consolidate compiler generated dependencies of target userver-uboost-coro
gmake[5]: выход из каталога «/home/user/svetit/src/back/auth/build_debug»
[  1%] Built target userver-pq-extra
gmake[5]: выход из каталога «/home/user/svetit/src/back/auth/build_debug»
[  1%] Built target create-session-key
[  2%] Built target userver-librseq
[  2%] Built target userver-http-parser
[  2%] Built target userver-llhttp
[  3%] Built target userver-uboost-coro
Consolidate compiler generated dependencies of target userver-universal
gmake[5]: выход из каталога «/home/user/svetit/src/back/auth/build_debug»
[ 20%] Built target userver-universal
gmake[5]: вход в каталог «/home/user/svetit/src/back/auth/build_debug»
Consolidate compiler generated dependencies of target userver-core
gmake[5]: выход из каталога «/home/user/svetit/src/back/auth/build_debug»
gmake[5]: вход в каталог «/home/user/svetit/src/back/auth/build_debug»
[ 20%] Building CXX object third_party/userver/core/CMakeFiles/userver-core.dir/src/engine/io/socket.cpp.o
[ 20%] Building CXX object third_party/userver/core/CMakeFiles/userver-core.dir/src/engine/io/pipe.cpp.o
[ 20%] Building CXX object third_party/userver/core/CMakeFiles/userver-core.dir/src/utils/userver_info.cpp.o
[ 20%] Linking CXX static library libuserver-cored.a
gmake[5]: выход из каталога «/home/user/svetit/src/back/auth/build_debug»
[ 85%] Built target userver-core
gmake[5]: вход в каталог «/home/user/svetit/src/back/auth/build_debug»
Consolidate compiler generated dependencies of target userver-postgresql
gmake[5]: выход из каталога «/home/user/svetit/src/back/auth/build_debug»
[ 94%] Built target userver-postgresql
gmake[5]: вход в каталог «/home/user/svetit/src/back/auth/build_debug»
Consolidate compiler generated dependencies of target svetit_auth_objs
gmake[5]: выход из каталога «/home/user/svetit/src/back/auth/build_debug»
gmake[5]: вход в каталог «/home/user/svetit/src/back/auth/build_debug»
[ 94%] Building CXX object CMakeFiles/svetit_auth_objs.dir/src/service/tokens_oidc.cpp.o
In file included from /home/user/svetit/src/back/auth/third_party/jwt-cpp/include/jwt-cpp/jwt.h:12,
                 from /home/user/svetit/src/back/auth/third_party/jwt-cpp/include/jwt-cpp/traits/kazuho-picojson/traits.h:12,
                 from /home/user/svetit/src/back/auth/src/service/tokens_oidc.cpp:16:
/home/user/svetit/src/back/auth/third_party/jwt-cpp/include/jwt-cpp/base.h: In static member function ‘static std::initializer_list<std::__cxx11::basic_string<char> > jwt::alphabet::helper::base64url_percent_encoding::fill()’:
/home/user/svetit/src/back/auth/third_party/jwt-cpp/include/jwt-cpp/base.h:87:48: warning: returning local ‘initializer_list’ variable ‘fill’ does not extend the lifetime of the underlying array [-Winit-list-lifetime]
   87 |                                         return fill;
      |                                                ^~~~
/home/user/svetit/src/back/auth/third_party/jwt-cpp/include/jwt-cpp/base.h:86:76: note: declared here
   86 |                                         std::initializer_list<std::string> fill{"%3D", "%3d"};
      |                                                                            ^~~~
In file included from /home/user/svetit/src/back/auth/third_party/jwt-cpp/include/jwt-cpp/traits/kazuho-picojson/traits.h:12,
                 from /home/user/svetit/src/back/auth/src/service/tokens_oidc.cpp:16:
/home/user/svetit/src/back/auth/third_party/jwt-cpp/include/jwt-cpp/jwt.h: In static member function ‘static std::string jwt::verify_ops::insensitive_string_claim<json_traits, in_header>::to_lower_unicode(const string&, const std::locale&)’:
/home/user/svetit/src/back/auth/third_party/jwt-cpp/include/jwt-cpp/jwt.h:3045:46: warning: ISO C++ forbids variable length array ‘mb’ [-Wvla]
 3045 |                                         char mb[MB_CUR_MAX];
      |                                              ^~
/home/user/svetit/src/back/auth/src/service/tokens_oidc.cpp: In member function ‘std::shared_ptr<jwt::verifier<jwt::default_clock, jwt::traits::kazuho_picojson> > svetit::auth::tokens::jwt_impl::GetVerifier(const string&, const string&)’:
/home/user/svetit/src/back/auth/src/service/tokens_oidc.cpp:41:54: error: could not convert ‘it.std::_Rb_tree_iterator<std::pair<const std::__cxx11::basic_string<char>, std::shared_ptr<jwt::verifier<jwt::default_clock, jwt::traits::kazuho_picojson> > > >::operator->()->std::pair<const std::__cxx11::basic_string<char>, std::shared_ptr<jwt::verifier<jwt::default_clock, jwt::traits::kazuho_picojson> > >::second.std::shared_ptr<jwt::verifier<jwt::default_clock, jwt::traits::kazuho_picojson> >::<anonymous>.std::__shared_ptr<jwt::verifier<jwt::default_clock, jwt::traits::kazuho_picojson>, __gnu_cxx::_S_atomic>::get()’ from ‘std::__shared_ptr<jwt::verifier<jwt::default_clock, jwt::traits::kazuho_picojson>, __gnu_cxx::_S_atomic>::element_type*’ {aka ‘jwt::verifier<jwt::default_clock, jwt::traits::kazuho_picojson>*’} to ‘std::shared_ptr<jwt::verifier<jwt::default_clock, jwt::traits::kazuho_picojson> >’
   41 |                                 return it->second.get();
      |                                        ~~~~~~~~~~~~~~^~
      |                                                      |
      |                                                      std::__shared_ptr<jwt::verifier<jwt::default_clock, jwt::traits::kazuho_picojson>, __gnu_cxx::_S_atomic>::element_type* {aka jwt::verifier<jwt::default_clock, jwt::traits::kazuho_picojson>*}
/home/user/svetit/src/back/auth/src/service/tokens_oidc.cpp:61:45: error: could not convert ‘res.std::pair<std::_Rb_tree_iterator<std::pair<const std::__cxx11::basic_string<char>, std::shared_ptr<jwt::verifier<jwt::default_clock, jwt::traits::kazuho_picojson> > > >, bool>::first.std::_Rb_tree_iterator<std::pair<const std::__cxx11::basic_string<char>, std::shared_ptr<jwt::verifier<jwt::default_clock, jwt::traits::kazuho_picojson> > > >::operator->()->std::pair<const std::__cxx11::basic_string<char>, std::shared_ptr<jwt::verifier<jwt::default_clock, jwt::traits::kazuho_picojson> > >::second.std::shared_ptr<jwt::verifier<jwt::default_clock, jwt::traits::kazuho_picojson> >::<anonymous>.std::__shared_ptr<jwt::verifier<jwt::default_clock, jwt::traits::kazuho_picojson>, __gnu_cxx::_S_atomic>::get()’ from ‘std::__shared_ptr<jwt::verifier<jwt::default_clock, jwt::traits::kazuho_picojson>, __gnu_cxx::_S_atomic>::element_type*’ {aka ‘jwt::verifier<jwt::default_clock, jwt::traits::kazuho_picojson>*’} to ‘std::shared_ptr<jwt::verifier<jwt::default_clock, jwt::traits::kazuho_picojson> >’
   61 |                 return res.first->second.get();
      |                        ~~~~~~~~~~~~~~~~~~~~~^~
      |                                             |
      |                                             std::__shared_ptr<jwt::verifier<jwt::default_clock, jwt::traits::kazuho_picojson>, __gnu_cxx::_S_atomic>::element_type* {aka jwt::verifier<jwt::default_clock, jwt::traits::kazuho_picojson>*}
At global scope:
cc1plus: note: unrecognized command-line option ‘-Wno-gnu-zero-variadic-macro-arguments’ may have been intended to silence earlier diagnostics
gmake[5]: *** [CMakeFiles/svetit_auth_objs.dir/build.make:412: CMakeFiles/svetit_auth_objs.dir/src/service/tokens_oidc.cpp.o] Ошибка 1
gmake[5]: выход из каталога «/home/user/svetit/src/back/auth/build_debug»
gmake[4]: *** [CMakeFiles/Makefile2:305: CMakeFiles/svetit_auth_objs.dir/all] Ошибка 2
gmake[4]: выход из каталога «/home/user/svetit/src/back/auth/build_debug»
gmake[3]: *** [CMakeFiles/Makefile2:347: CMakeFiles/svetit_auth.dir/rule] Ошибка 2
gmake[3]: выход из каталога «/home/user/svetit/src/back/auth/build_debug»
gmake[2]: *** [Makefile:192: svetit_auth] Ошибка 2
gmake[2]: выход из каталога «/home/user/svetit/src/back/auth/build_debug»
make[1]: *** [Makefile:33: build-debug] Ошибка 2
make[1]: выход из каталога «/home/user/svetit/src/back/auth»
make: *** [Makefile:39: run-bin-auth] Ошибка 2
```